### ISSUE TITLE
Implement optional downsampling as part of preprocessing

### ISSUE THREAD

**Issue by jbarlow83** [jbarlow83/OCRmyPDF#443]
(As requested by @tcurdt)

Some users want optional downsampling of images, but we don't want to do this all the time in case the file has high resolution scans (e.g. film).

Like Acrobat, should downsample images above some threshold.

TBD: 
* What to do if we don't have any reason to flatten the PDF to images (no deskew, clean) - we should probably downsample on an image by image basis, rather than page by page, in this case, which suggests putting downsampling in the optimizer
* Generally it makes sense to downsample any image that would be too big Tesseract, because at least then we can do something with it
* Test image before OCR and downsample images that are too big and then scale them up
* To a degree it is better to aggressively lower JPEG quality than downsample, because JPEG encoders will save the bit budget for high detail areas. This starts to break down for very large images because they take up a lot of memory when decompressed and make the file unwieldy. So we may want to satisfy a request to downsample with some combination of downsampling and JPEG quality lowering.

**Comment by tcurdt** [comment]
@heinrich-ulbricht My use case is about reducing the embedded image size - but not necessarily completely replacing them. But the outcomes are pretty much aligned. Archiving a PDF with OCR data and reduced image data.

**Comment by jbarlow83** [comment]
@heinrich-ulbricht You are correct that what you are describing is not currently possible with OCRmyPDF.

**Comment by heinrich-ulbricht** [comment]
I created a ticket specific to my use case to not highjack this ticket further.

**Comment by jbarlow83** [comment]
ocrmypdf now picks  the output resolution more carefully, so this is partially done.
